lib/TPP/Transforms/HoistVectorTransfers.cpp

```diff
@@ -0,0 +1,209 @@
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_HOISTVECTORTRANSFERS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace vector;
+
+namespace mlir {
+namespace tpp {
+
+struct HoistVectorTransferOp : OpRewritePattern<vector::ContractionOp> {
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractOp,
+                                PatternRewriter &rewriter) const override {
+
+    // Check the vector contract operation satisfies the required pattern.
+    // Check the ACC, LHS, and RHS of contract operation
+    auto vectorReadOpAcc =
+        contractOp.getAcc().getDefiningOp<mlir::vector::TransferReadOp>();
+    if (!vectorReadOpAcc)
+      return failure();
+
+    auto subviewOp =
+        vectorReadOpAcc.getOperand(0).getDefiningOp<memref::SubViewOp>();
+    if (!subviewOp)
+      return failure();
+
+    auto vectorReadOpLHS =
+        contractOp.getLhs().getDefiningOp<mlir::vector::TransferReadOp>();
+    if (!vectorReadOpLHS)
+      return failure();
+
+    auto vectorReadOpRHS =
+        contractOp.getRhs().getDefiningOp<mlir::vector::TransferReadOp>();
+    if (!vectorReadOpRHS)
+      return failure();
+
+    // Check the operation type MatMul, B-MatMul, or BR-MatMul
+    auto contractIteratorTypes = contractOp.getIteratorTypesArray();
+    size_t reductionCount = 0;
+    for (size_t i = 0; i < contractIteratorTypes.size(); i++) {
+      if (contractIteratorTypes[i] == vector::IteratorType::reduction)
+        reductionCount++;
+    }
+
+    auto vectorReadOpLHSType = cast<ShapedType>(vectorReadOpLHS.getType());
+    auto vectorReadOpRHSRank =
+        (cast<ShapedType>(vectorReadOpRHS.getType())).getRank();
+
+    if (reductionCount == 2 &&
+        (vectorReadOpLHSType.getRank() != 3 || vectorReadOpRHSRank != 3))
+      return failure();
+
+    if (reductionCount == 1)
+      return rewriter.notifyMatchFailure(
+          contractOp, "Batch matmul operation not supported yet");
+
+    if (reductionCount > 2)
+      return rewriter.notifyMatchFailure(
+          contractOp, "The vector contract operation is not a gemm");
+
+    // Check the K-dim to be 1
+    int64_t K =
+        vectorReadOpLHSType.getDimSize(vectorReadOpLHSType.getRank() - 1);
+    if (K != 1)
+      return rewriter.notifyMatchFailure(contractOp, "K dim is not 1");
+
+    // Check whether the linalg tiling + vector contract pattern matches for the
+    // 4-nested loop structure
+    auto oldKForOp = contractOp->getParentOfType<scf::ForOp>();
+    if (!oldKForOp)
+      return rewriter.notifyMatchFailure(
+          contractOp, "Not a linalg tile + vector contract operation");
+
+    auto oldReductionForOp = oldKForOp->getParentOfType<scf::ForOp>();
+    if (!oldReductionForOp)
+      return rewriter.notifyMatchFailure(
+          contractOp, "Not a linalg tile + vector contract operation");
```

Comment:
repetitive notification message doesn't help identify what wen't wrong, when it does. Follow my main comment strategy on error messages.

Reply:
Yes. Updated the pass with functions as said in the main comment.